src/adt/print/tree.jl

```diff
@@ -136,20 +135,118 @@ function (p::Printer)(node)
     end
 end
 
+function children(node)
+    error("unimplemented children method for $(typeof(node))")
+end
 function print_node(io::IO, node)
     error("unimplemented print_node method for $(typeof(node))")
+end    
+function print_child_annotation(io::IO, node, child, annotation)
+    error("unimplemented print_child_annotation method for ($(typeof(node)), $(typeof(child)), $(typeof(annotation)))")
 end
 
-function children(node)
-    error("unimplemented children method for $(typeof(node))")
+should_print_annotation(ch) = applicable(keys, ch)
+should_print_annotation(::AbstractVector) = false
+should_print_annotation(::Tuple) = false
+should_print_annotation(::Base.Generator) = false
+
+end # module Multiline
+
+
+module Inline
+
+Base.@kwdef mutable struct State
+    depth::Int = 0
+    precedence::Int = 0 # precedence of the parent expression
+end
+
+Base.@kwdef struct Color
+    delimiter::Symbol = :light_red
+    brackets::Vector{Symbol} = [:yellow, :green, :cyan, :magenta, :light_gray, :default, :light_red]
+end
+
+struct Printer{IO_t <: IO}
+    io::IO_t
+    color::Color
+    state::State
 end
 
-should_printkeys(ch) = applicable(keys, ch)
-should_printkeys(::AbstractVector) = false
-should_printkeys(::Tuple) = false
-should_printkeys(::Base.Generator) = false
+function Printer(io::IO; color::Color = Color())
+    Printer(io, color, State())
+end
+
+function (p::Printer)(node)
+    c = p.color
+    print(xs...) = Base.print(p.io, xs...)
+    printstyled(xs...;kw...) = Base.printstyled(p.io, xs...; kw...)
+    print_node(node) = Inline.print_node(p.io, node)
+    print_node_suffix(node) = Inline.print_node_suffix(p.io, node)
+    print_child_annotation(node, child, annotation) = Inline.print_child_annotation(p.io, node, child, annotation)
+    print_child_annotation_suffix(node, child, annotation) = Inline.print_child_annotation_suffix(p.io, node, child, annotation)
+
+    function join(xs, delimiter, should_print_annotation = false)
+        for (i, x) in enumerate(xs)
+            if should_print_annotation
+                child, annotation = x
+                print_child_annotation(node, child, annotation)
+                p(child)
+                print_child_annotation_suffix(node, child, annotation)
+            else
+                p(x)
+            end
+            i < length(xs) && printstyled(delimiter; color = c.delimiter)
+        end
+    end
+
+    function bracket(f, left, right)
```

Comment:
I believe bracket means something else, this should be `parenthesis` or `paren` as abbr

Reply:
I guess `bracket` would be better here because we could have different brackets including `(), [], {}, <>` while parenthesis only refers to `()`.